Fix s390x regressions

uweigand opened the pull request:
- Add relocation handling needed after PR #3275
- Fix incorrect handling of signed constants detected by PR #3056 test
- Disable fuzzing tests that require pre-built v8 binaries
- Disable cranelift test that depends on i128
- Temporarily disable memory64 tests

<!--

Please ensure that the following steps are all taken care of before submitting
the PR.

- [ ] This has been discussed in issue #..., or if not, please tell us why
  here.
- [ ] A short description of what this does, why it is needed; if the
  description becomes long, the matter should probably be discussed in an issue
  first.
- [ ] This PR contains test cases, if meaningful.
- [ ] A reviewer from the core maintainer team has been assigned for this PR.
  If you don't know who could review this, please indicate so. The list of
  suggested reviewers on the right can help you.

Please ensure all communication adheres to the [code of
conduct](https://github.com/bytecodealliance/wasmtime/blob/master/CODE_OF_CONDUCT.md).
-->

uweigand commented:
Note: In addition to this PR, the endian fix in PR #3329 and the rsix build fix https://github.com/bytecodealliance/rsix/pull/53 are required to make the CI tests fully pass again on s390x.

uweigand commented:
Also: Using qemu 6.1 we should now be able to turn on upstream CI for the s390x platform.

sunfishcode commented:
I opened https://github.com/bytecodealliance/wasmtime/pull/3331 to pull in the s390x fix in `rsix`.

uweigand commented:
> I opened #3331 to pull in the s390x fix in `rsix`.

Thanks, @sunfishcode !

alexcrichton commented on build.rs:
```diff
@@ -201,6 +201,8 @@ fn ignore(testsuite: &str, testname: &str, strategy: &str) -> bool {
             // No simd support yet for s390x.
             ("simd", _) if platform_is_s390x() => return true,
+            // No memory64 support yet for s390x.
+            ("memory64", _) if platform_is_s390x() => return true,
```
How come memory64 isn't implemented on s390x? Does it lower to unimplemented clif things at the moment?

uweigand replied:
Yes.   This is all fixable, but I'd prefer to do that separately.

alexcrichton commented on cranelift/codegen/src/isa/s390x/inst/mod.rs:
```diff
@@ -3621,6 +3624,7 @@ impl MachInstLabelUse for LabelUse {
             // This can address any valid CodeOffset.
             LabelUse::BranchRIL => 0x7fff_ffff,
             LabelUse::PCRel32 => 0x7fff_ffff,
+            LabelUse::PCRel32Dbl => 0x7fff_ffff,
```
Given the shift, is the range actually one bit larger than what's specified here?

uweigand replied:
The range is actually +/- 4GB, not +/-2 GB.   But that apparently cannot be represented as CodeOffset anyway, so I simply accept all valid CodeOffsets here.

alexcrichton replied:
Could this return `0xffff_ffff` as a placeholder? One day I think it'd be good to represent some code offsets in this area as a 64-bit integer, but that I think is a much bigger change.

uweigand replied:
I tried that when adding BranchRIL (which has the same range) initially, and it caused failures (some kind of overflows in common code), so that's what I ended up with.   I guess this could be revisited at some point, but then we should do it for all affected LabelUse types, not just PCRel32Dbl.

alexcrichton replied:
Can the overflow be fixed with a `saturating_add` or something like that somewhere? If not can the comment be updated that this should be 4GB but it's specified as 2GB for now?

uweigand replied:
That would probably help, I'll give it a try.   It's still better to do that in a separate PR I think.

alexcrichton reviewed (commented):
I think someone else will want to review the `cranelift/codegen/src/isa/s390x/lower.rs` bits, but all the other bits look good to me.

uweigand commented:
@alexcrichton you were right, I just had to add a single `saturating_add`.    Given that this is just a one-line obvious fix to common code,  I've now updated this PR after all.   Let me know if you'd prefer me to take it out again.

Note that BranchRI max ranges were just completely wrong, but that type isn't currently used anywhere so it didn't matter.  I've now fixed this and used precise max ranges for all types, only capped by a max of 0xffff_ffff due to the u32 type.

alexcrichton commented:
Nah yeah that's what I would have expected for a `saturating_add` and it looks good to me! I think someone will still want to look over the sconst changes to the s390x backend though.

afonso360 commented:
> Also: Using qemu 6.1 we should now be able to turn on upstream CI for the s390x platform.

Just want to mention that we have now upgraded to 6.1 in #3325

uweigand commented:
> > Also: Using qemu 6.1 we should now be able to turn on upstream CI for the s390x platform.
> 
> Just want to mention that we have now upgraded to 6.1 in #3325

Excellent.   Not sure exactly what we need to do to add a CI job for s390x, I haven't done much with GitHub actions before ...

alexcrichton commented:
To add a CI job theoretically all that needs to be done is to copy/paste [this block](https://github.com/bytecodealliance/wasmtime/blob/d20194fa4c45c68ff1fea315eabc262e854c3074/.github/workflows/main.yml#L230-L237) to add another entry below it, and then update the various values for s390x instead of aarch64.

uweigand commented:
> To add a CI job theoretically all that needs to be done is to copy/paste [this block](https://github.com/bytecodealliance/wasmtime/blob/d20194fa4c45c68ff1fea315eabc262e854c3074/.github/workflows/main.yml#L230-L237) to add another entry below it, and then update the various values for s390x instead of aarch64.

Thanks!  I'll give it a try once the two PRs are merged.

cfallin reviewed (approved):
> Thanks! @cfallin do you want to have a look at those?

The label/MachBuffer changes look good to me!

uweigand commented:
So can this be merged now?   This is now the final bit blocking successful builds on s390x ...

uweigand commented:
> To add a CI job theoretically all that needs to be done is to copy/paste [this block](https://github.com/bytecodealliance/wasmtime/blob/d20194fa4c45c68ff1fea315eabc262e854c3074/.github/workflows/main.yml#L230-L237) to add another entry below it, and then update the various values for s390x instead of aarch64.

This is now https://github.com/bytecodealliance/wasmtime/pull/3372 .   I also had to disable the lightbeam tests (just like on aarch64).
